keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,626 @@
+- If dual-stack endpoint address pairs are present, then each pair should be printed on its own line as show below.
+Example:
+```
+       kube-master# kubectl describe service nginx-service
+       .
+       .
+       .
+       Endpoints:   {[fd00:db8:1::2]:80,192.168.1.3:80}
+                    {[fd00:db8:2::2]:80,192.168.2.3:80}
+       .
+       .
+       .
+       kube-master#
+```
+
+### kube-proxy Operation
+
+Kube-proxy will be modified to drive iptables and ip6tables in parallel. This will require the implementation of a second "proxier" interface in the Kube-Proxy server in order to modify and track changes to both tables. This is required in order to allow exposing services via both IPv4 and IPv6, e.g. using Kubernetes:
+  - NodePort
+  - ExternalIPs
+
+#### Kube-Proxy Startup Configuration Changes
+
+##### Multiple bind addresses configuration
+The existing "--bind-address" option for the [kube-proxy startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/) will be modified to support multiple IP addresses in a comma-separated list (rather than a single IP string).
+```
+  --bind-address  stringSlice   (IP addresses, in a comma separated list, Default: [0.0.0.0,])
+```
+Only the first address of each IP family will be used; all others will be ignored.
+
+##### Multiple cluster CIDRs configuration
+The existing "--cluster-cidr" option for the [kube-proxy startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/) will be modified to support multiple IP CIDRs in a comma-separated list (rather than a single IP CIDR).
+A new [kube-proxy configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/) argument will be added to allow a user to specify multiple cluster CIDRs.
+```
+  --cluster-cidr  ipNetSlice   (IP CIDRs, in a comma separated list, Default: [])
+```
+Only the first CIDR for each IP family will be used; all others will be ignored.
+
+### IPVS Support and Operation
+
+Since IPVS functionality does not yet include IPv6 support (see [cloudnativelabs/kube-router Issue #307](https://github.com/cloudnativelabs/kube-router/issues/307)), support for IPVS functionality in a dual-stack cluster is considered a "nice-to-have" or stretch goal.
+
+### Health/Liveness/Readiness Probes for Dual-Stack Pods
+
+Currently, health, liveness, and readiness probes are defined without any concern for IP addresses or families. For the first release of dual-stack support, a cluster administrator will be able to select the preferred IP family to use for probes when a pod has both IPv4 and IPv6 addresses. For this selection, a new "--preferred-probe-ip-family" argument for the for the [kubelet startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/) will be added:
+```
+  --preferred-probe-ip-family  string   ["ipv4", "ipv6", or "none". Default: "none", meaning use the pod's default IP]
```

Comment:
@mtaufen should we be adding flags or JUST component config fields?

Reply:
resolved in https://github.com/kubernetes/enhancements/pull/808. It was removed as we updated the behavior here.
> - For health/liveness/readiness probe support, the default behavior will not change and an additional optional field would be added to the pod specification and is respected by kubelet. This will allow application developers to select a preferred IP family to use for implementing probes on dual-stack pods.